### PR TITLE
Smiley backend for futuhours

### PR DESCRIPTION
**Pull request by mixman**

### CODE REVIEW

File: smileys-app/LICENSE [hunk 1]
```diff
@@ -0,0 +1,30 @@
+Copyright Jussi Vaihia (c) 2017
```
Comment (phadej) [inline smileys-app/LICENSE]: Futurice

File: smileys-app/src/Futurice/App/Smileys/Logic.hs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+getSmileys
+    :: (MonadIO m, MonadBaseControl IO m)
+    => Ctx
+    -> Maybe FUM.UserName
+    -> Maybe Day
+    -> Maybe Day
+    -> m SmileysReport
+getSmileys ctx _ start end = do
+  now <- liftIO $ getCurrentTime
```
Comment (phadej) [inline smileys-app/src/Futurice/App/Smileys/Logic.hs]: `currentTime` from `MonadTime`

Comment (phadej) [inline smileys-app/src/Futurice/App/Smileys/Logic.hs]: ... or direclty `today <- currentDay`

File: smileys-app/src/Futurice/App/Smileys/Logic.hs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+
+postSmiley
+    :: (MonadIO m, MonadBaseControl IO m, MonadError ServantErr m)
+    => Ctx
+    -> Maybe FUM.UserName
+    -> PostSmiley
+    -> m Res
+postSmiley ctx mfum req = do
+  withResource (ctxPostgresPool ctx) $ \conn -> do
```
Comment (phadej) [inline smileys-app/src/Futurice/App/Smileys/Logic.hs]: 4 spaces

File: smileys-app/src/Futurice/App/Smileys/Logic.hs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+               , "VALUES (?, ?, ?, ?) ON CONFLICT (username, day) DO UPDATE"
+               , "SET entries = EXCLUDED.entries, smiley = EXCLUDED.smiley"
+               ]
+        let smiley_entries = textShow $ _postSmileyEntries req
+        let smiley_day = textShow $ _postSmileyDate req
+        let smiley_user = FUM._getUserName fumUsername
+        let smiley_state = textShow $ _postSmileySmiley req
+        _ <- liftIO $ Postgres.execute conn
+               insertQuery
+               [ smiley_entries, smiley_user, smiley_state, smiley_day]
```
Comment (phadej) [inline smileys-app/src/Futurice/App/Smileys/Logic.hs]: use tuple, not list.

File: smileys-app/src/Futurice/App/Smileys/Types.hs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  , _postSmileySmiley   :: !SmileyState
+  } deriving (Eq, Ord, Show, Typeable, Generic)
+
+data Res = Res
+    { _resStatus :: !Text
+    } deriving (Show)
+
+type ProjectId = Int
+type TaskId = Int
+type SmileyState = Int
```
Comment (phadej) [inline smileys-app/src/Futurice/App/Smileys/Types.hs]: `SmileyValue`?

File: smileys-app/src/Futurice/App/Smileys/Types.hs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+-- TODO: Naming: Smiley? SmileyEntries?
+data Smiley = Smiley
```
Comment (phadej) [inline smileys-app/src/Futurice/App/Smileys/Types.hs]: `SmileyTask`? (project id is redundant, add a comment why it's good idea to have it there still).

File: smileys-app/src/Futurice/App/Smileys/Types.hs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+instance ToSchema Smileys where declareNamedSchema = sopDeclareNamedSchema
+
+-------------------------------------------------------------------------------
+-- Reports
+-------------------------------------------------------------------------------
+
+type SmileysReport = Report
+    "Smileys report"
+    ReportGenerated
+    (Vector :$ NP I '[Text, Text, Int, Day])
```
Comment (phadej) [inline smileys-app/src/Futurice/App/Smileys/Types.hs]: I'd write a special type, e.g. `SmileyEntry` or `SmileyReportEntry` with right fields, see e.g. https://github.com/futurice/haskell-mega-repo/blob/99bcbbb30fedb93cd632754b0ac218807c301c3a/reports-app/src/Futurice/App/Reports/MissingHours.hs#L53

File: smileys-app/src/Futurice/App/Smileys/Types.hs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+-- Smiley
+-------------------------------------------------------------------------------
+
+data PostSmiley = PostSmiley
+  { _postSmileyEntries :: ![Smiley]
+  , _postSmileyDate :: !Day
+  , _postSmileySmiley   :: !SmileyState
+  } deriving (Eq, Ord, Show, Typeable, Generic)
+
+data Res = Res
```
Comment (phadej) [inline smileys-app/src/Futurice/App/Smileys/Types.hs]: very non-descriptive name...

File: tajna.yaml [hunk 9]
```diff
@@ -81,12 +81,13 @@ envs:
...
     GITHUBPROXY_PORT:   "8811"
     FUMAPP_PORT:        "8812"
+    SMILEYS_PORT:       "8813"
 
     AVATARAPP_EKGPORT:     "9901"
     CONTACTSAPI_EKGPORT:   "9902"
     # These two are the same on purpose
     FUTUHOURSAPI_EKGPORT:  "9903"
-    FUTUHOURSMOCK_EKPORT:  "9903"
+    FUTUHOURSMOCK_EKGPORT:  "9903"
```
Comment (phadej) [inline tajna.yaml]: remove space to align port numbers :)

File: smileys-app/src/Futurice/App/Smileys/Types.hs (hunk 8 above)
Comment (phadej) [inline smileys-app/src/Futurice/App/Smileys/Types.hs]: ... maybe `PostResult` ?

File: smileys-app/src/Futurice/App/Smileys/Logic.hs [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+               , "VALUES (?, ?, ?, ?) ON CONFLICT (username, day) DO UPDATE"
+               , "SET entries = EXCLUDED.entries, smiley = EXCLUDED.smiley"
+               ]
+        let smiley_entries = _postSmileyEntries req
+        let smiley_day = textShow $ _postSmileyDate req
+        let smiley_user = FUM._getUserName fumUsername
+        let smiley_state = textShow $ _postSmileySmiley req
+        _ <- liftIO $ Postgres.execute conn
+               insertQuery
+               ( smiley_entries, smiley_user, smiley_state, smiley_day)
```
Comment (phadej) [inline smileys-app/src/Futurice/App/Smileys/Logic.hs]: more specifically `Vector` inserted as *Postgres Array*: http://hackage.haskell.org/package/postgresql-simple-0.5.3.0/docs/src/Database-PostgreSQL-Simple-ToField.html#line-285, but as the field is `Text` the default conversion is applied there.

**Comment by mixman** [comment]
Ok to merge?

File: smileys-app/schema.psql [hunk 11]
```diff
@@ -0,0 +1,12 @@
+CREATE SCHEMA smileys;
+
+CREATE TABLE smileys.trail (
+  id SERIAL PRIMARY KEY,
+  entries TEXT[] NOT NULL,
```
Comment (phadej) [inline smileys-app/schema.psql]: should this be `JSON[]`

File: smileys-app/schema.psql [hunk 12]
```diff
@@ -0,0 +1,12 @@
+CREATE SCHEMA smileys;
+
+CREATE TABLE smileys.trail (
+  id SERIAL PRIMARY KEY,
+  entries JSON NOT NULL,
```
Comment (phadej) [inline smileys-app/schema.psql]: does it work? i.e. `JSON`, not `JSON[]` ?

File: smileys-app/smileys-app.cabal [hunk 13]
```diff
@@ -0,0 +1,76 @@
+-- This file has been generated from package.yaml by hpack version 0.17.0.
+--
+-- see: https://github.com/sol/hpack
+
+name:           smileys-app
+version:        0
+synopsis:       Proxy conf management
```
Comment (phadej) [inline smileys-app/smileys-app.cabal]: not true :)

File: smileys-app/package.yaml [hunk 14]
```diff
@@ -0,0 +1,61 @@
+name: smileys-app
+version: "0"
+synopsis: Proxy conf management
```
Comment (phadej) [inline smileys-app/package.yaml]: not true :)

File: smileys-app/src/Futurice/App/Smileys.hs [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+-- Smileys modules
+import Futurice.App.Smileys.API
+import Futurice.App.Smileys.Config (Config (..))
+import Futurice.App.Smileys.Ctx
+import Futurice.App.Smileys.Logic
+
+server :: Ctx -> Server SmileysAPI
+server ctx = pure "smileys backend"
+    :<|> (postSmiley ctx)
```
Comment (phadej) [inline smileys-app/src/Futurice/App/Smileys.hs]: unnecessary parentheses

File: smileys-app/src/Futurice/App/Smileys/Logic.hs [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+    -> Maybe FUM.UserName
+    -> PostSmiley
+    -> m Res
+postSmiley ctx mfum req = do
+    withResource (ctxPostgresPool ctx) $ \conn -> do
+        mcase (mfum <|> ctxMockUser ctx) (throwError err403) $ \fumUsername -> do
+            let insertQuery = fromString $ unwords $
+                 [ "INSERT INTO smileys.trail as c (entries, username, smiley, day)"
+                 , "VALUES (?, ?, ?, ?) ON CONFLICT (username, day) DO UPDATE"
+                 , "SET entries = CAST(EXCLUDED.entries AS json), smiley = EXCLUDED.smiley"
```
Comment (phadej) [inline smileys-app/src/Futurice/App/Smileys/Logic.hs]: this cast is smelly

File: smileys-app/src/Futurice/App/Smileys/Logic.hs [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+                 ]
+            let smiley_entries = _postSmileyEntries req
+            let smiley_day = textShow $ _postSmileyDate req
+            let smiley_user = FUM._getUserName fumUsername
+            let smiley_state = textShow $ _postSmileySmiley req
+            _ <- liftIO $ Postgres.execute conn
+                     insertQuery
+                     ( smiley_entries, smiley_user, smiley_state, smiley_day)
```
Comment (phadej) [inline smileys-app/src/Futurice/App/Smileys/Logic.hs]: maybe write `ToRow` instance for `PostSmiley`?

File: smileys-app/src/Futurice/App/Smileys/Logic.hs [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+                (s, e)
+
+postSmiley
+    :: (MonadIO m, MonadBaseControl IO m, MonadError ServantErr m)
+    => Ctx
+    -> Maybe FUM.UserName
+    -> PostSmiley
+    -> m Res
+postSmiley ctx mfum req = do
+    withResource (ctxPostgresPool ctx) $ \conn -> do
```
Comment (phadej) [inline smileys-app/src/Futurice/App/Smileys/Logic.hs]: I'd do the connection acquisition after we check for the mock user.

File: stack-ghc-8.0.yaml [hunk 19]
```diff
@@ -41,6 +41,7 @@ packages:
 - servant-Chart/
 - servant-binary-tagged/
 - servant-status/
+- smileys-app/
```
Comment (phadej) [inline stack-ghc-8.0.yaml]: should add to `stack-lts-6.yaml` too; also do `stack clean; stack build --pedantic` to get the errors. (See travis)

**Comment by phadej** [comment]
LGTM. I'll try locally later today.

**Pull request merged**